src/compile/mark/rule.ts

```diff
@@ -12,25 +12,48 @@ export namespace rule {
     let p: any = {};
 
     // TODO: support explicit value
+    if(model.config().mark.orient === 'vertical') {
+      if (model.has(X)) {
+        p.x = position(model, X);
+      } else {
+        p.x = { value : 0};
+      }
 
-    // vertical
-    if (model.has(X)) {
-      p.x = position(model, X);
+      if (model.has(Y)) {
+        p.y = position(model, Y);
+      } else {
+        p.y = { field: { group: 'height' } };
+      }
 
-      p.y = { value: 0 };
-      p.y2 = {
-          field: {group: 'height'}
+      if (model.has(Y2)) {
+        p.y2 = {
+          scale: model.scaleName(Y),
+          field: model.field(Y2)
         };
-    }
+      } else {
+        p.y2 = { field: { group: 'height' } };
+      }
+    } else {
+      if (model.has(Y)) {
+        p.y = position(model, Y);
+      } else {
+        p.y = { value: 0 };
+      }
 
-    // horizontal
-    if (model.has(Y)) {
-      p.y = position(model, Y);
+      if (model.has(X)) {
+        p.x = position(model, X);
+      } else {
+        p.x = { value: 0 };
+      }
 
-      p.x = { value: 0 };
-      p.x2 = {
-          field: {group: 'width'}
+      if (model.has(X2)) {
+        p.x2 = {
```

Comment:
At first I was using different scales, I think I forgot to fix that. Although, I agree to get rid of `position()` because of the reasons you mentioned and its more clear with just specifying it there.